### PR TITLE
Added utc date format (RFC 3339) for json formatted logging

### PR DESCRIPTION
**Pull request by xogoodnow**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added date_time_utc parameter to configs so json log format supports UTC date time (RFC 3339/ISO8601)

Related issue:
https://github.com/ClickHouse/ClickHouse/issues/64853

- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

By default logs are shown in UTC format (RFC 3339) but when json log "date_time" parameter in json log formatting config block is enabled in server configuration file, the logs are only shown in Unix format. The added parameter ("date_time_utc") solves this issue.

- Sample Log Output: 
- Before : 
{"date_time":"1730388956.627566","thread_name":"","thread_id":"253130","level":"Debug","query_id":"","logger_name":"Application","message":"Shut down storages.","source_file":"programs\/server\/Server.cpp; auto DB::Server::main(const std::vector<std::string> &)::(anonymous class)::operator()() const","source_line":"968"}

- Now:
{"date_time_utc":"2024-10-31T15:20:50Z.281688","date_time":"1730388050.281688","thread_name":"","thread_id":"251834","level":"Trace","query_id":"","logger_name":"BackgroundSchedulePool\/BgMBSchPool","message":"Waiting for threads to finish.","source_file":"src\/Core\/BackgroundSchedulePool.cpp; DB::BackgroundSchedulePool::~BackgroundSchedulePool()","source_line":"222"}

-->

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Loggers/OwnJSONPatternFormatter.cpp [hunk 1]
```diff
@@ -62,6 +69,26 @@ void OwnJSONPatternFormatter::formatExtended(const DB::ExtendedLogMessage & msg_
...
     DB::writeChar('{', wb);
 
+    if (!date_time_utc.empty())
+    {
+        writeJSONString(date_time_utc, wb, settings);
+        DB::writeChar(':', wb);
+
+        DB::writeChar('\"', wb);
+        static const DateLUTImpl & utc_time_zone = DateLUT::instance("UTC");
+        /// Change delimiters in date for compatibility with old logs.
```
Comment (vdimir) [inline src/Loggers/OwnJSONPatternFormatter.cpp:79]: Could you please clarify - change delimiters from what to what? And compatibility with which old logs?

**Review by vdimir — APPROVED** [review]

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 1996fbf31dd00975cb0c0bfec53511df95574428 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71314/1996fbf31dd00975cb0c0bfec53511df95574428/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71306/33cbc540d523888eea630f467718ac84f723f068/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: programs/server/config.xml [hunk 2]
```diff
@@ -70,12 +70,15 @@
         You can specify log format(for now, JSON only). In that case, the console log will be printed
         in specified format like JSON.
         For example, as below:
+
         {"date_time":"1650918987.180175","thread_name":"#1","thread_id":"254545","level":"Trace","query_id":"","logger_name":"BaseDaemon","message":"Received signal 2","source_file":"../base/daemon/BaseDaemon.cpp; virtual void SignalListener::run()","source_line":"192"}
+        {"date_time_utc":"2022-04-26T20:36:27+00:00","thread_name":"#1","thread_id":"254545","level":"Trace","query_id":"","logger_name":"BaseDaemon","message":"Received signal 2","source_file":"../base/daemon/BaseDaemon.cpp; virtual void SignalListener::run()","source_line":"192"}
```
Comment (janeklb) [inline programs/server/config.xml:75]: What is the actual format that will be produced? The PR's body suggests 
```json
{"date_time_utc":"2024-10-31T15:20:50Z.281688", 
```
But here we have 
```json
{"date_time_utc":"2022-04-26T20:36:27+00:00",
```

Comment (janeklb) [reply programs/server/config.xml:75]: Ah, looks like the implementation points at the former `"2024-10-31T15:20:50Z.281688"`

File: src/Loggers/OwnJSONPatternFormatter.cpp [hunk 3]
```diff
@@ -62,6 +69,26 @@ void OwnJSONPatternFormatter::formatExtended(const DB::ExtendedLogMessage & msg_
...
+        writeDateTimeTextISO(msg_ext.time_seconds, 0, wb, utc_time_zone);
+        DB::writeChar('.', wb);
+        DB::writeChar('0' + ((msg_ext.time_microseconds / 100000) % 10), wb);
+        DB::writeChar('0' + ((msg_ext.time_microseconds / 10000) % 10), wb);
+        DB::writeChar('0' + ((msg_ext.time_microseconds / 1000) % 10), wb);
+        DB::writeChar('0' + ((msg_ext.time_microseconds / 100) % 10), wb);
+        DB::writeChar('0' + ((msg_ext.time_microseconds / 10) % 10), wb);
+        DB::writeChar('0' + ((msg_ext.time_microseconds / 1) % 10), wb);
```
Comment (janeklb) [inline src/Loggers/OwnJSONPatternFormatter.cpp:87]: Unfortunately this does not produce RFC 3339 compliant datetime strings. The "parts of second" should appear _before_ the `Z`. (see https://it-tools.tech/date-converter)

eg. it should be `2024-10-31T15:20:50.281688Z` (not `2024-10-31T15:20:50Z.281688`)


If this is a limitation of `writeDateTimeTextISO` then I would suggest dropping the sub-second precision from this PR, and just logging it the second-level for now (eg. `2024-10-31T15:20:50Z`)

**Comment by janeklb** [comment]
Wonderful! This functionality is sorely needed for ingesting ClickHouse logs

File: src/Loggers/OwnJSONPatternFormatter.cpp (hunk 3 above)
Comment (janeklb) [reply src/Loggers/OwnJSONPatternFormatter.cpp:87]: maybe `writeDateTimeText` can be used?

Comment (xogoodnow) [reply src/Loggers/OwnJSONPatternFormatter.cpp]: Thank you for pointing this out.
Fixed the issue.
Now: "2024-11-04T19:01:54Z"
Technically "writeDateTimeText" could be used as well but this is 8601 (RFC3339).

**Comment by xogoodnow** [comment]
Also fixed the issue with style check

**Comment by xogoodnow** [comment]
Hi @janeklb, 

I do not understand the issue with the pipeline.
Could you please tell me what is the problem?

Cheers

File: programs/server/config.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (xogoodnow) [reply programs/server/config.xml:75]: Sure thing.

**Comment by xogoodnow** [comment]
@vdimir, there was a faulty pull request to my branch, I attempted to rewrite the history for the commit and I made a mistake in the process.
I will close this PR and submit the changed in another request
I apologize 
Cheers

**Comment by xogoodnow** [comment]
here is the new PR
https://github.com/ClickHouse/ClickHouse/pull/71560

PS: sorry for the inconvenience

**Comment by vdimir** [comment]
@xogoodnow No worries. You could force push new commits to this branch and PR would be updated, but creating new PR is also fine!

**Comment by xogoodnow** [comment]
It became too much of a mess. So I opened a new PR.
Thank you @vdimir

**Pull request closed without merging**